src/main/java/com/couchbase/lite/router/Router.java

```diff
@@ -1887,19 +1875,46 @@ private RevisionInternal update(Database _db, String docID, Body body, boolean d
 
         RevisionInternal result = null;
         try {
-            synchronized (_db) {
-                if (isLocalDoc) {
-                    result = _db.getStore().putLocalRevision(rev, prevRevID, true);
-                } else {
-                    result = _db.putRevision(rev, prevRevID, allowConflict);
+            if (isLocalDoc) {
+                // NOTE: putLocalRevision() does not use transaction internally with obeyMVCC=true
+
+                final Database fDb = _db;
+                final RevisionInternal _rev = rev;
+                final String _prevRevID = prevRevID;
+                final List<RevisionInternal> _revs = new ArrayList<RevisionInternal>();
+                try {
+                    fDb.getStore().runInTransaction(new TransactionalTask() {
+                        @Override
+                        public boolean run() {
+                            try {
+                                RevisionInternal r = fDb.getStore().putLocalRevision(_rev, _prevRevID, true);
+                                _revs.add(r);
+                                return true;
+                            } catch (CouchbaseLiteException e) {
+                                throw new RuntimeException(e);
+                            }
+                        }
+                    });
+                    // success
+                    if (_revs.size() > 0)
+                        result = _revs.get(0);
+                } catch (RuntimeException ex) {
+                    if (ex.getCause() != null &&
+                            ex.getCause().getCause() != null &&
+                            ex.getCause().getCause() instanceof CouchbaseLiteException)
+                        throw (CouchbaseLiteException) ex.getCause().getCause();
+                    else
+                        throw new CouchbaseLiteException(ex, Status.INTERNAL_SERVER_ERROR);
                 }
+            } else {
+                // putRevision() uses transaction internally.
+                result = _db.putRevision(rev, prevRevID, allowConflict);
```

Comment:
[add()](https://github.com/couchbase/couchbase-lite-java-core/blob/a35361bc47ecb6c55849694deece22fa4aaf7c62/src/main/java/com/couchbase/lite/store/SQLiteStore.java#L1336) uses transaction in store layer.